node/src/components/consensus/highway_core/state.rs

```diff
@@ -566,8 +575,54 @@ impl<C: Context> State<C> {
                 return Err(VoteError::ValueAfterTerminalBlock);
             }
         }
-        // TODO: Validate against LNC.
-        Ok(())
+        match self.validate_lnc(wvote) {
+            Ok(_) => Ok(()),
+            Err(LncError::MissingEndorsement(vhash)) => {
+                panic!("Protocol state is missing endorsement for {:?}", &vhash)
+            }
+            Err(LncError::NaiveCitation(vidx)) => Err(VoteError::LncNaiveCitation(vidx)),
+        }
+    }
+
+    /// Validates whether `wvote` violates the LNC rule.
+    /// Returns the first vote's hash that should have been endorsed but wasn't.
+    ///
+    /// Vote violates LNC rule if it cites naively an equivocation.
+    /// If it cites equivocator then it needs to endorse votes that cite equivocating votes.
+    fn validate_lnc(&self, wvote: &WireVote<C>) -> Result<(), LncError<C>> {
+        let mut need_endorsement = wvote.panorama.need_endorsements(self);
+        let key_hashes: Vec<C::Hash> = need_endorsement.keys().cloned().collect();
+        // Remove keys this vote claims are endorsed.
+        for v in key_hashes {
+            if wvote.is_endorsed(&v) {
+                if !self.is_endorsed(&v) {
```

Comment:
This part should be done in `missing_dependency`.

Reply:
This is a sanity check, because we do download it in `missing_dependency`. I can remove it.